README
======
README
======

:Last update:   Date: 2016-04-30
:Revision:      2.4.1
:Contact:       Brian Dolbec <dolsen@gentoo.org>


About
-----

layman is the Gentoo overlay manager and allows to integrate
experimental software packages into the main distribution. It can also
be used as a manager for version control repositories.


Links
-----

Primary project web site:

 - http://layman.sourceforge.net/

Project site on SourceForge:

 - http://sourceforge.net/projects/layman

Project summary on Ohloh:

 - http://ohloh.net/projects/layman/

Project summary on Freshmeat:

 - http://freshmeat.net/projects/layman/

Bug tracker:

 - http://bugs.gentoo.org

Wiki:

 - http://gentoo-wiki.com/Portage_Overlay_Listing#Layman

Combined RSS feed:

 - http://www.google.com/reader/shared/user/02645926629531261525/label/%5Bproject%5D%20layman

Git repository:
 - http://git.overlays.gentoo.org/gitweb/?p=proj/layman.git;a=summary


File description
----------------

AUTHORS      : The developers of this project
ChangeLog    : A detailed change log
CHANGES      : A summary of the project changes
COPYING      : Licensing information
HACKING      : Developers notes for the project
INSTALL      : Installation instructions
MANIFEST.in  : The listing of files that get included into the python
               package.
README       : This file.
README.api   : Example api use information
RELEASE_NOTES: Release notes for the latest version
TODO         : Task list for the project


CHANGES
CHANGES
-------

Version 2.4.3 - Release 2020-04-24
==================================
 - Bug 613936 Drop portage_native_kwargs
 - Bug 674594 typo in laymansync.pylayman.py
 - Fix python 3.7 and python 3.8 compatibility
 - several more typo fixes, added more debug statements...

Version 2.4.2 - Release 2017-02-01
==================================
 - Bug 587474 Fix a missed eprefix placeholder for repos_conf setting.
 - Bug 539336 Adds exception handling
 - Fix commit 63808ef00f94 mis-applied 'sourcefile' attribute changes

Version 2.4.1 - Release 2016-05-06
==================================
 - adds sourcefile to plug-in modules, fixes bug #582154

Version 2.4.0 - Release 2016-05-03
==================================
 - adds support for different overlay database types
 - adds json database support
 - adds sqlite database support
 - adds database migration tool in layman-updater
 - adds AddDeleteDB tests in test suite
 - adds file locker utility class
 - corrects importing of laymansync classes
 - corrects type in enable/disable_repos()
 - adds multiple owner support
 - deprecates get_all_info() function
 - adds repos.conf dir creation if it doesn't exist
 - fixes bug #540012
 - moves Layman class initialization into a layman sub-file
 - renames layman.py to Layman.py
 - corrects misspelling of "separate"
 - changes default conf_type to "repos.conf"
 - changes arg group "Additions Options" to "Path Options"
 - adds protocol filtering for Overlay.{add, update}()
 - adds binary mode aware compatibility to fileopen()
 - adds sanity checking for repos_conf location
 - fixes Unicode() external test

Version 2.3.0 - Release 2015-02-08
==================================
 - Renamed portage sync plugin to laymansync
 - adds portage sync plugin to setup.py for installation
 - Completely removes doctest remnants
 - adds fix that quells initial warning if layman does not see an install.xml file
 - adds spelling correction on external.py header
 - modifies remotedb.py to reflect ssl-fetch changes
 - fixes misuse of .keys() in maker.py causing py3.x compat issues
 - whitespace cleanup in maker.py
 - adds skipping of protocol verification if protocol guessing fails in maker.py
 - fixes undefined xml error in maker.py
 - adds checks for which archive overlay types to test in external test suite
 - converts all assertEquals calls to assertEqual (assertEquals will be deprecated)
 - modifies and improves testing method for external test suite
 - renames self.overlay to self.overlays in reposconf.py
 - modifies output message encoding to become indiscriminate on type
 - changes ssl-fetch kwargs-info level to 5 to limit verbosity
 - modifies checking for sync plugin name to reflect new name in reposconf.py
 - layman-updater updates for better repos.conf migration


Version 2.2.0 - Release 2014-09-28
==================================
 - man page updates
 - Portage sync plug-in to sync layman overlays via portage
 - layman-overlay-maker utility to create overlay definitions
 - Multiple description field support
 - reloading config function to get local overlay definitions
 - output attribute to OptionConfig
 - branch and layman-type info to repos.conf configs
 - sync-type checking for repos.conf configs
 - New --storage cli option to specify where you want an overlay installed
 - fixes bug when overlay isn't in the remote list
 - New --disable and --enable cli option to globally disable a locally
   installed overlay
 - Centeralizes archive overlay type code
 - Changes clean_tar config option to clean_archive
 - fixes hard-coded defaults and EPREFIXIFIES repos_conf config option
 - fixes format string in _verify_overlay_type function
 - renames --setup_help cli option to --setup-help
 - adds support for multiple cli option args
 - adds interactive session for unofficial overlay installations
 - migrates docstring tests to external test-suite
 - adds layman-mounter utility handle overlay mounting
 - adds squashfs overlay support
 - adds squashfs overlay external tests
 - adds modular plug-in system
 - makes overlay types modular plug-ins
 - makes repo config types modular plug-ins
 - removes g-common overlay support
 - fixes output use in proxies()
 - fixes default command line value overriding of overlay_defs and
   storage config options
 - fixes improper tracking of operation results for Overlay.update()
 - adds use of repos.conf module plug-in when running layman-updater
 - adds check in repos.conf to write all overlays to a blank file
 - fixes improper import of make.conf module in updater.py
 - fixes closing of iterable file extension list in tar.py


Version 2.1.0 - Released 2014-07-23
===================================
 - man page updates
 - python 2.7 and python 3.x compatibility
 - migrate from optparse to argparse
 - Support for repos.conf configuration files
 - RepoConfManager to handle all repo configuration types
 - URL updating for outdated source URLs
 - Command line option to re-add an overlay
 - Overlay type checking to ensure the overlay type of an
   installed overlay is the same as the overlay type reported
   by the remote database
 - IRC and branch tag support for overlay xmls
 - clean_tar config option to clean up local tar overlay files


Version 2.0.0 final - Released 2012-12-16
=========================================
 - re-released rc5 as 2.0.0 final


Version 2.0.0_rc5 - Released 2012-11-19
=======================================
 - fix a regression in calling read_config()
 - fix layman-updater when creating a new make.conf file
 - fix truncating of long overlay names
 - remove a mispelled duplicate program file
 - auto-detect portage's make.conf location


Version 2.0.0_rc4 - Released 2012-10-20
=======================================
 - man page updates, formatting changes
 - fix unknown variable
 - fix bug 384085, deprecated "bzr get" to "bzr branch"
 - py2, py3 compatability changes so 2to3 will work correctly (unfinished)
 - make the tests more py3 compatibile

action.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#################################################################################
# LAYMAN ACTIONS
#################################################################################
# File:       action.py
#
#             Handles layman actions.
#
# Copyright:
#             (c) 2005 - 2006 Gunnar Wrobel
#             Distributed under the terms of the GNU General Public License v2
#
# Author(s):
#             Gunnar Wrobel <wrobel@gentoo.org>
#
''' Provides the different actions that can be performed by layman.'''

__version__ = "$Id: action.py 312 2007-04-09 19:45:49Z wrobel $"

#===============================================================================
#
# Dependencies
#
#-------------------------------------------------------------------------------

import sys

from   layman.db                import DB, RemoteDB

from   layman.debug             import OUT

#===============================================================================
#
# Class Fetch
#
#-------------------------------------------------------------------------------

class Fetch:
    ''' Fetches the overlay listing.

    >>> import os
    >>> here = os.path.dirname(os.path.realpath(__file__))
    >>> cache = os.tmpnam()
    >>> config = {'overlays' :
    ...           'file://' + here + '/tests/testfiles/global-overlays.xml',
    ...           'cache' : cache,
    ...           'nocheck'    : True,
    ...           'proxy' : None,
    ...           'quietness':3}
    >>> a = Fetch(config)
    >>> a.run()
    0
    >>> b = open(a.db.path(config['overlays']))
    >>> b.readlines()[24]
    '      A collection of ebuilds from Gunnar Wrobel [wrobel@gentoo.org].\\n'

    >>> b.close()
    >>> os.unlink(a.db.path(config['overlays']))

    >>> a.db.overlays.keys()
    [u'wrobel', u'wrobel-stable']
    '''

    def __init__(self, config):
        self.db = RemoteDB(config)

    def run(self):
        '''Fetch the overlay listing.'''
        try:
            self.db.cache()
        except Exception, error:
            OUT.die('Failed to fetch overlay list!\nError was: '
                    + str(error))

        return 0

#===============================================================================
#
# Class Sync
#
#-------------------------------------------------------------------------------

class Sync:
    ''' Syncs the selected overlays.'''

    def __init__(self, config):

        self.db = DB(config)

        self.rdb = RemoteDB(config)

        self.selection = config['sync']

        enc = sys.getfilesystemencoding()
        self.selection = [i.decode(enc) for i in self.selection]

        if config['sync_all'] or 'ALL' in self.selection:
            self.selection = self.db.overlays.keys()

    def run(self):
        '''Synchronize the overlays.'''

        OUT.debug('Updating selected overlays', 6)

        warnings = []
        success  = []
        for i in self.selection:
            ordb = self.rdb.select(i)
            odb = self.db.select(i)
            if ordb and odb and ordb.src != odb.src:
                warnings.append(
                    'The source of the overlay "' + i + '" seems to have c'
                    'hanged. You currently sync from "' + odb.src + '" whi'
                    'le the global layman list reports "' + ordb.src + '" '
                    'as correct location. Please consider removing and rea'
                    'dding the overlay!')

            try:
                self.db.sync(i)
                success.append('Successfully synchronized overlay "' + i + '".')
            except Exception, error:
                warnings.append(
                    'Failed to sync overlay "' + i + '".\nError was: '
                    + str(error))

        if success:
            OUT.info('\nSuccess:\n------\n', 3)
            for i in success:
                OUT.info(i, 3)
                
        if warnings:
            OUT.warn('\nErrors:\n------\n', 2)
            for i in warnings:
                OUT.warn(i + '\n', 2)
            return 1

        return 0

#===============================================================================
#
# Class Add
#
#-------------------------------------------------------------------------------

class Add:
    ''' Adds the selected overlays.'''

    def __init__(self, config):

        self.config = config

        self.db = DB(config)

        self.rdb = RemoteDB(config)

        self.selection = config['add']

        enc = sys.getfilesystemencoding()
        self.selection = [i.decode(enc) for i in self.selection]

        if 'ALL' in self.selection:
            self.selection = self.rdb.overlays.keys()

    def run(self):
        '''Add the overlay.'''

        OUT.debug('Adding selected overlays', 6)

        result = 0

        for i in self.selection:
            overlay = self.rdb.select(i)

            OUT.debug('Selected overlay', 7)

            if overlay:
                try:
                    self.db.add(overlay)
                    OUT.info('Successfully added overlay "' + i + '".', 2)
                except Exception, error:
                    OUT.warn('Failed to add overlay "' + i + '".\nError was: '
                             + str(error), 2)
                    result = 1
            else:
                OUT.warn('Overlay "' + i + '" does not exist!', 2)
                result = 1

        return result

#===============================================================================
#
# Class Delete
#
#-------------------------------------------------------------------------------

class Delete:
    ''' Deletes the selected overlays.'''

    def __init__(self, config):

        self.db = DB(config)

        self.selection = config['delete']

        enc = sys.getfilesystemencoding()
        self.selection = [i.decode(enc) for i in self.selection]

        if 'ALL' in self.selection:
            self.selection = self.db.overlays.keys()

    def run(self):
        '''Delete the overlay.'''

        OUT.debug('Deleting selected overlays', 6)

        result = 0

        for i in self.selection:
            overlay = self.db.select(i)

            OUT.debug('Selected overlay', 7)

            if overlay:
                try:
                    self.db.delete(overlay)
                    OUT.info('Successfully deleted overlay "' + i + '".', 2)
                except Exception, error:
                    OUT.warn('Failed to delete overlay "' + i + '".\nError was: '
                             + str(error), 2)
                    result = 1
            else:
                OUT.warn('Overlay "' + i + '" does not exist!', 2)
                result = 1

        return result

#===============================================================================
#
# Class Info
#
#-------------------------------------------------------------------------------

class Info:
    ''' Print information about the specified overlays.

    >>> import os
    >>> here = os.path.dirname(os.path.realpath(__file__))
    >>> cache = os.tmpnam()
    >>> config = {'overlays' :
    ...           'file://' + here + '/tests/testfiles/global-overlays.xml',
    ...           'cache'  : cache,
    ...           'proxy'  : None,
    ...           'info'   : ['wrobel'],
    ...           'nocheck'    : False,
    ...           'verbose': False,
    ...           'quietness':3}
    >>> a = Info(config)
    >>> a.rdb.cache()
    >>> OUT.color_off()
    >>> a.run()
    * wrobel
    * ~~~~~~
    * Source  : https://overlays.gentoo.org/svn/dev/wrobel
    * Contact : nobody@gentoo.org
    * Type    : Subversion; Priority: 10
    * 
    * Description:
    *   Test
    * 
    0
    '''

    def __init__(self, config):

        OUT.debug('Creating RemoteDB handler', 6)

        self.rdb    = RemoteDB(config)
        self.config = config

        self.selection = config['info']

        enc = sys.getfilesystemencoding()
        self.selection = [i.decode(enc) for i in self.selection]

        if 'ALL' in self.selection:
            self.selection = self.rdb.overlays.keys()

    def run(self):
        ''' Print information about the selected overlays.'''

        for i in self.selection:
            overlay = self.rdb.select(i)
            # Is the overlay supported?
            OUT.info(overlay.__str__(), 1)
            if not overlay.is_official():
                OUT.warn('*** This is no official gentoo overlay ***\n', 1)
            if not overlay.is_supported():
                OUT.error('*** You are lacking the necessary tools to install t'
                          'his overlay ***\n')

        return 0

#===============================================================================
#
# Class List
#
#-------------------------------------------------------------------------------

class List:
    ''' Lists the available overlays.

    >>> import os
    >>> here = os.path.dirname(os.path.realpath(__file__))
    >>> cache = os.tmpnam()
    >>> config = {'overlays' :
    ...           'file://' + here + '/tests/testfiles/global-overlays.xml',
    ...           'cache'  : cache,
    ...           'proxy'  : None,
    ...           'nocheck'    : False,
    ...           'verbose': False,
    ...           'quietness':3}
    >>> a = List(config)
    >>> a.rdb.cache()
    >>> OUT.color_off()
    >>> a.run()
    * wrobel                    [Subversion] (source: https://overlays.gentoo.or...)
    0
    >>> a.config['verbose'] = True
    >>> a.run()
    * wrobel
    * ~~~~~~
    * Source  : https://overlays.gentoo.org/svn/dev/wrobel
    * Contact : nobody@gentoo.org
    * Type    : Subversion; Priority: 10
    * 
    * Description:
    *   Test
    * 
    * *** This is no official gentoo overlay ***
    * 
    * wrobel-stable
    * ~~~~~~~~~~~~~
    * Source  : rsync://gunnarwrobel.de/wrobel-stable
    * Contact : nobody@gentoo.org
    * Type    : Rsync; Priority: 50
    * 
    * Description:
    *   A collection of ebuilds from Gunnar Wrobel [wrobel@gentoo.org].
    * 
    0
    '''

    def __init__(self, config):

        OUT.debug('Creating RemoteDB handler', 6)

        self.rdb    = RemoteDB(config)
        self.config = config

    def run(self):
        ''' List the available overlays.'''

        for i in self.rdb.list(self.config['verbose']):
            # Is the overlay supported?
            if i[1]:
                # Is this an official overlay?
                if i[2]:
                    OUT.info(i[0], 1)
                # Unofficial overlays will only be listed if we are not
                # checking or listing verbose
                elif self.config['nocheck'] or self.config['verbose']:
                    # Give a reason why this is marked yellow if it is a verbose
                    # listing
                    if self.config['verbose']:
                        OUT.warn('*** This is no official gentoo overlay ***\n', 1)
                    OUT.warn(i[0], 1)
            # Unsupported overlays will only be listed if we are not checking
            # or listing verbose
            elif self.config['nocheck'] or self.config['verbose']:
                # Give a reason why this is marked red if it is a verbose
                # listing
                if self.config['verbose']:
                    OUT.error('*** You are lacking the necessary tools to insta'
                              'll this overlay ***\n')
                OUT.error(i[0])

        return 0

#===============================================================================
#
# Class ListLocal
#
#-------------------------------------------------------------------------------

class ListLocal:
    ''' Lists the local overlays.'''

    def __init__(self, config):
        self.db = DB(config)
        self.config = config

    def run(self):
        '''List the overlays.'''

        for i in self.db.list(self.config['verbose']):

            OUT.debug('Printing local overlay.', 8)

            # Is the overlay supported?
            if i[1]:
                # Is this an official overlay?
                if i[2]:
                    OUT.info(i[0], 1)
                # Unofficial overlays will only be listed if we are not
                # checking or listing verbose
                else:
                    # Give a reason why this is marked yellow if it is a verbose
                    # listing
                    if self.config['verbose']:
                        OUT.warn('*** This is no official gentoo overlay ***\n', 1)
                    OUT.warn(i[0], 1)
            # Unsupported overlays will only be listed if we are not checking
            # or listing verbose
            else:
                # Give a reason why this is marked red if it is a verbose
                # listing
                if self.config['verbose']:
                    OUT.error('*** You are lacking the necessary tools to insta'
                              'll this overlay ***\n')
                OUT.error(i[0])

        return 0

#===============================================================================
#
# Class Actions
#
#-------------------------------------------------------------------------------

class Actions:
    '''Dispatches to the actions the user selected. '''

    # Given in order of precedence
    actions = [('fetch',      Fetch),
               ('add',        Add),
               ('sync',       Sync),
               ('info',       Info),
               ('sync_all',   Sync),
               ('delete',     Delete),
               ('list',       List),
               ('list_local', ListLocal),]

    def __init__(self, config):

        # Make fetching the overlay list a default action
        if not 'nofetch' in config.keys():
            # Actions that implicitely call the fetch operation before
            fetch_actions = ['fetch', 'sync', 'sync_all', 'list']
            for i in fetch_actions:
                if i in config.keys():
                    # Implicitely call fetch, break loop
                    Fetch(config).run()
                    break

        result = 0

        for i in self.actions:

            OUT.debug('Checking for action', 7)

            if i[0] in config.keys():
                result += i[1](config).run()


        if not result:
            sys.exit(0)
        else:
            sys.exit(1)
            
#===============================================================================
#
# Testing
#
#-------------------------------------------------------------------------------

if __name__ == '__main__':
    import doctest, sys

    # Ignore warnings here. We are just testing
    from warnings     import filterwarnings, resetwarnings
    filterwarnings('ignore')

    doctest.testmod(sys.modules[__name__])

    resetwarnings()
